pkg/backends/calico/calico_test.go

```diff
@@ -0,0 +1,146 @@
+package calico
+
+import (
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+func addEndpointNode(ep *v1.Endpoints, nodename string) {
+	ep.Subsets = append(ep.Subsets, v1.EndpointSubset{
+		Addresses: []v1.EndpointAddress{
+			v1.EndpointAddress{
+				NodeName: &nodename}}})
+}
+
+var _ = Describe("RouteGenerator", func() {
+	var (
+		rg   *routeGenerator
+		meta metav1.ObjectMeta
+		svc  *v1.Service
+		ep   *v1.Endpoints
+	)
+
+	BeforeEach(func() {
+		rg = &routeGenerator{
+			nodeName:    "foobar",
+			svcIndexer:  cache.NewIndexer(cache.MetaNamespaceKeyFunc, nil),
+			epIndexer:   cache.NewIndexer(cache.MetaNamespaceKeyFunc, nil),
+			svcRouteMap: make(map[string]string),
+			client: &client{
+				cache: make(map[string]string),
+			},
+		}
+		meta = metav1.ObjectMeta{Namespace: "foo", Name: "bar"}
```

Comment:
I'm not a huge fan of defining these sorts of objects in BeforeEach.

It sort of implies that we're always going to use the same objects, but some tests might use different services / endpoints / etc. I'd prefer a function like `buildSimpleService` or something that returns them and can be called by each test if needed.

Reply:
Or, at least move into a more specific BeforeEach() rather than Before _every_ route generator test.

Reply:
yeah... the balance between writing DRY vs expressive tests...
i tried to define the top level objects as generically as possible but the inner tests do end up mutating them ever so slightly (which is the whole point of `addEndpointSubset`)
interesting suggestion on defining a factory method to generate the objects...

Reply:
Yep, I generally find DRY to be limited in value in many situations, so if there's a tradeoff like this one I'd fall on the "expressive tests" side of things.

We can get both by restructuring a little bit. But I think `BeforeEach()` should include only functionality that is required before each test.